src/query-builder/SelectQueryBuilder.ts

```diff
@@ -1903,6 +1912,56 @@ export class SelectQueryBuilder<Entity extends ObjectLiteral>
         }
     }
 
+    private lazyCount(entitiesAndRaw: { entities: Entity[]; raw: any[] }) {
+        const hasLimit =
+            this.expressionMap.limit !== undefined &&
+            this.expressionMap.limit !== null
+        if (this.expressionMap.joinAttributes.length > 0 && hasLimit) {
+            return undefined
+        }
+
+        let maxResults = undefined
+
+        const hasTake =
+            this.expressionMap.take !== undefined &&
+            this.expressionMap.take !== null
+        if (hasTake) {
+            maxResults = this.expressionMap.take
+        }
+
+        if (hasLimit) {
+            // limit overrides take when no join is defined
+            maxResults = this.expressionMap.limit
+        }
+
+        if (
+            maxResults !== undefined &&
+            entitiesAndRaw.entities.length == maxResults
+        ) {
+            return undefined
+        }
+
+        let previousResults = 0
+        if (
+            this.expressionMap.skip !== undefined &&
+            this.expressionMap.skip !== null &&
+            this.expressionMap.skip > 0
+        ) {
+            previousResults = this.expressionMap.skip
+        }
+
+        if (
+            this.expressionMap.offset !== undefined &&
+            this.expressionMap.offset !== null &&
+            this.expressionMap.offset > 0
+        ) {
+            // offset overrides skip when no join is defined
+            previousResults = this.expressionMap.offset
+        }
```

Comment:
```suggestion
        const hasSkip = this.expressionMap.skip !== undefined &&
            this.expressionMap.skip !== null &&
            this.expressionMap.skip > 0
        const hasOffset = this.expressionMap.offset !== undefined &&
            this.expressionMap.offset !== null &&
            this.expressionMap.offset > 0
        // offset overrides skip when no join is defined
        const previousResults = hasOffset
            ? this.expressionMap.offset
            : hasSkip
            ? this.expressionMap.skip
            : 0;
```

Reply:
I added type assertion to make typescript happy with the undefined that could actually not happen. I'm not 100% happy with that, if you have a better idea, feel free to share!